scripts/benchmarking/train_gaussian_denoiser.py

```diff
@@ -0,0 +1,247 @@
+        "commit_hash": commit_hash,
+    }
+elif params["model"] == "gs_drunet":
+    model_params = LIONParameter(
+        in_channels=1,
+        out_channels=1,
+        int_channels=params["channels"],
+        kernel_size=(params["kernel_size"], params["kernel_size"]),
+        n_blocks=params["n_blocks"],
+        use_noise_level=False,
+        bias_free=params["bias_free"],
+        act="elu",
+        enforce_positivity=params["enforce_positivity"],
+    )
+    model = GSDRUNet(model_params).cuda()
+    config = {
+        "n_blocks": params["n_blocks"],
+        "int_channels": params["channels"],
+        "kernel_size": params["kernel_size"],
+        "bias_free": params["bias_free"],
+        "lr": params["lr"],
+        "epochs": params["epochs"],
+        "noise_level": params["noise_level"],
+        "commit_hash": commit_hash,
+    }
+else:
+    raise NotImplementedError(f"Model {params['model']} has not been implemented!")
+
+experiment_id = uuid.uuid1()
+experiment_name = f"{params['model']}"
+if params["bias_free"]:
+    experiment_name += "_bias_free"
+if "dncnn" in params["model"]:
+    experiment_name += f"_depth={params['depth']}"
+else:
+    experiment_name += f"_n_blocks={params['n_blocks']}"
+experiment_name += f"_noise_level={params['noise_level']}_{experiment_id}"
+print(experiment_name)
+print(config)
+wandb.init(project="benchmarking_ct", config=config, name=experiment_name)
+
+optimiser = torch.optim.Adam(model.parameters(), lr=params["lr"], betas=(0.9, 0.9))
+random_crop = RandomCrop((256, 256))
+random_erasing = RandomErasing()
+experiment = ct_benchmarking.GroundTruthCT()
+
+training_data = experiment.get_training_dataset()
+validation_data = experiment.get_validation_dataset()
+testing_data = experiment.get_testing_dataset()
+
+print(
+    f"N_train={len(training_data)}, N_val={len(validation_data)}, N_test={len(testing_data)}"
+)
+
+batch_size = 1
+
+training_dataloader = torch.utils.data.DataLoader(
+    training_data, batch_size, shuffle=True
+)
+validation_dataloader = torch.utils.data.DataLoader(
+    validation_data, batch_size, shuffle=False
+)
+testing_dataloader = torch.utils.data.DataLoader(
+    testing_data, batch_size, shuffle=False
+)
+
+with open("normalisation.json", "r") as fp:
+    normalisation = json.load(fp)
+    x_min, x_max = normalisation["x_min"], normalisation["x_max"]
+
+best_val_psnr = -inf
+
+losses = []
+val_psnrs = []
+for epoch in range(params["epochs"]):
+    model.train()
+    for it, x in enumerate(training_dataloader):
+        x = x.cuda()
+        x = (x - x_min) / (x_max - x_min)
+        patches = random_erasing(torch.cat([random_crop(x) for _ in range(5)], dim=0))
+        optimiser.zero_grad()
+        y = patches + params["noise_level"] * torch.randn_like(patches)
+        recon = model(y)
+        loss = torch.mean((recon - patches) ** 2)
+        loss.backward()
+        grad_norm = mean_grad_norm(model)
+        losses.append(loss.item())
+        optimiser.step()
```

Comment:
So you erase some patches, then step on the loss and then test on the patches+nose? Or `random_erasing` gives you the patches??

You did this because of memory, right?

Reply:
`random_crop` selects random square patches from the image, while `random_erasing` erases random rectangles from the patches (intended to counteract overfitting). After applying those transforms, I add noise and  pass the noisy cropped+erased patches through the denoiser to train. Training on patches was necessary at least for the gradient-step denoisers because of memory constraints.